core/templates/dev/head/pages/exploration_player/PlayerServices.js

```diff
@@ -407,6 +422,28 @@ oppia.factory('oppiaPlayerService', [
           deferred.resolve(DEFAULT_PROFILE_IMAGE_PATH);
         }
         return deferred.promise;
+      },
+      recordSolutionHit: function() {
+        // Check that the current card is an active card by checking whether
+        // the active card's solution is the same as the current solution.
```

Comment:
This doesn't seem right to me. The caller should pass the state name corresponding to the solution that's being hit to this function; this function should then check that state name against the "active" one. State names for different states are guaranteed to be unique. Solutions for different states are not.

(In fact, this check could probably go in the caller. That way, this function is only called when we are sure we want to record the event -- it's not a "maybe record".)

Reply:
@seanlip I just noticed that in the caller, the state name is always playerPositionService.getCurrentStateName(). So, this method is always called only with an active state name. So, I don't think there is any point in even doing this check. I'm removing it for now. What do you think?

Reply:
Well, I would suggest testing manually and seeing what happens when you view a solution for a state that you previously answered (both in the cases where you've just given a correct answer, and also in the case where you navigate backwards through the cards to view a solution to a card from long ago). If the event doesn't trigger in those two cases, but does trigger in the case where you click the solution button before submitting the correct answer, then your code is probably correct.

Reply:
So, the TutorCard viewSolution method retrieves the solution for the active card. I'm not 100% sure about this - but I don't think solutions can be viewed for cards that you progress back to through progress dots (The code that retrieves solutions always seems to use the active card's solution - so probably doesn't occur for non-active cards).
So, recording solutions seems correct now.

Reply:
Please try to verify so you're 100% sure -- is there even a "View Solution" button when you navigate back through progress dots? Also /cc @kevintab95 who coded this and might be able to provide advice.

It would be odd if the active card's solution was accessible through clicking "View Solution" on a previous card, btw. That is unexpected behaviour and is a bug that should be fixed; your code should not rely on such behaviour.

Reply:
@seanlip Just confirmed that with an exploration from oppia.org . I tried using all hints and finally got the solution and progressed to the next state. When I came back through progress dots, couldn't use hints. Since you can't submit answers when you traverse back through progress dots, doesn't make sense to have solutions either right?

Reply:
Don't make assumptions -- create a local exploration with some hints and a solution, answer all hints so that the "view solution" button is showing, go to the next state, then go back and check if you can still view the solution in the old state. Verifying this is something you need to do if you want to be sure you're developing this correctly.

Reply:
@seanlip. Done. Checked it. View solution button is not there when you traverse back through dots.